File: cases/taint/cpp/interprocedural-exception-persistence-positive/interprocedural_exception_persistence_positive.cpp
struct FlowException {};

struct FlowBox {
    int value = 0;
};

int dfb_source() { // DFB-SOURCE: exception-persistence-input
    return 1;
}

void dfb_sink(int value) {} // DFB-SINK: exception-persistence-sink

[[noreturn]] void store_and_throw(FlowBox &box, int value) {
    box.value = value; // DFB-WITNESS: exception-persistence-store
    throw FlowException{}; // DFB-WITNESS: exception-persistence-throw
}

int recover(FlowBox &box, int value) {
    try {
        store_and_throw(box, value);
    } catch (const FlowException &) {
        return box.value; // DFB-WITNESS: exception-persistence-recovery
    }
    return -1;
}

void run() {
    FlowBox box;
    dfb_sink(recover(box, dfb_source()));
}

File: cases/taint/php/interprocedural-exception-persistence-positive/interprocedural_exception_persistence_positive.php
<?php
function dfb_source(): string { // DFB-SOURCE: exception-persistence-input
    return "tainted";
}

function dfb_sink(string $value): void {} // DFB-SINK: exception-persistence-sink

final class FlowException extends \RuntimeException
{
}

class FlowBox
{
    public string $value;

    public function __construct(string $value)
    {
        $this->value = $value;
    }
}

function store_and_throw(FlowBox $box, string $value): void
{
    $box->value = $value; // DFB-WITNESS: exception-persistence-store
    throw new FlowException("exceptional exit"); // DFB-WITNESS: exception-persistence-throw
}

function recover(FlowBox $box, string $value): string
{
    try {
        store_and_throw($box, $value);
        return "unreachable";
    } catch (FlowException $_ignored) {
        return $box->value; // DFB-WITNESS: exception-persistence-recovery
    }
}

function run(): void
{
    dfb_sink(recover(new FlowBox("seed"), dfb_source()));
}
